Samples/BackgroundBlur/BackgroundBlur/TransformBlur.cpp

```diff
@@ -0,0 +1,1550 @@
+#include "TransformBlur.h"
+
+
+#define CHECK_HR(hr) if (FAILED(hr)) { goto done; }
+// Video FOURCC codes.
+const FOURCC FOURCC_YUY2 = MAKEFOURCC('Y', 'U', 'Y', '2');
+const FOURCC FOURCC_UYVY = MAKEFOURCC('U', 'Y', 'V', 'Y');
+const FOURCC FOURCC_NV12 = MAKEFOURCC('N', 'V', '1', '2');
+const FOURCC FOURCC_RGB24 = 20;
+const FOURCC FOURCC_RGB32 = 22;
+
+// Static array of media types (preferred and accepted).
+const GUID* g_MediaSubtypes[] =
+{
+    &MFVideoFormat_RGB32,
+    //&MEDIASUBTYPE_NV12,
+    //&MEDIASUBTYPE_YUY2,
+    //&MEDIASUBTYPE_UYVY
+};
+
+// Number of media types in the aray.
+DWORD g_cNumSubtypes = ARRAY_SIZE(g_MediaSubtypes);
+
+// GetImageSize: Returns the size of a video frame, in bytes.
+HRESULT GetImageSize(FOURCC fcc, UINT32 width, UINT32 height, DWORD* pcbImage);
+// ConvertMFTypeToDXVAType: Convert an IMFMediaType to DXVA_Desc
+HRESULT ConvertMFTypeToDXVAType(IMFMediaType * pType, DXVA2_VideoDesc * pDesc);
+
+//-------------------------------------------------------------------
+// Name: CreateInstance
+// Description: Static method to create an instance of the source.
+//
+// pUnkOuter:   Aggregating object or NULL.
+// iid:         IID of the requested interface on the source.
+// ppSource:    Receives a ref-counted pointer to the source.
+//-------------------------------------------------------------------
+HRESULT TransformBlur::CreateInstance(IUnknown* pUnkOuter, REFIID iid, void** ppMFT)
+{
+    if (ppMFT == NULL)
+    {
+        return E_POINTER;
+    }
+
+    // This object does not support aggregation.
+    if (pUnkOuter != NULL)
+    {
+        return CLASS_E_NOAGGREGATION;
+    }
+
+    HRESULT hr = S_OK;
+    TransformBlur* pMFT = new TransformBlur(hr);
+    if (pMFT == NULL)
+    {
+        CHECK_HR(hr = E_OUTOFMEMORY);
+    }
+
+    CHECK_HR(hr = pMFT->QueryInterface(iid, ppMFT));
+
+done:
+    SAFE_RELEASE(pMFT);
+    return hr;
+}
+
+// Constructor
+TransformBlur::TransformBlur(HRESULT& hr) :
+    m_nRefCount(1),
+    m_pSample(NULL),
+    m_pInputType(NULL),
+    m_pOutputType(NULL),
+    m_videoFOURCC(0),
+    m_imageWidthInPixels(0),
+    m_imageHeightInPixels(0),
+    m_cbImageSize(0),
+    m_pD3DDeviceManager(NULL),
+    m_pHandle(NULL),
+    m_pD3DDevice(NULL),
+    m_pD3DVideoDevice(NULL)
+{
+}
+
+//Destructor
+TransformBlur::~TransformBlur()
+{
+    assert(m_nRefCount == 0);
+
+    SAFE_RELEASE(m_pInputType);
+    SAFE_RELEASE(m_pOutputType);
+    SAFE_RELEASE(m_pSample);
+
+    m_pD3DDeviceManager->CloseDeviceHandle(m_pHandle);
+}
+
+// IUnknown methods
+ULONG TransformBlur::AddRef()
+{
+    return InterlockedIncrement(&m_nRefCount);
+}
+
+ULONG TransformBlur::Release()
+{
+    ULONG uCount = InterlockedDecrement(&m_nRefCount);
+    if (uCount == 0)
+    {
+        delete this;
+    }
+    // For thread safety, return a temporary variable.
+    return uCount;
+}
+
```

Comment:
Isn't there a default implementation of these if you inherit from IUnknown? Just trying to understand why you need to refcount manually.

Also, do we need e.g. copy constructors to update the refcount? It feels like this implementation has gaps

Reply:
(Or make it not copyable)